File: Algorithms/429_N-ary_Tree_Level_Order_Traversal/solution.py
from collections import defaultdict
from typing import *

"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""


class Solution:
    def levelOrder(self, root: 'Node') -> List[List[int]]:
        if not root:
            return []
        else:
            q: List[Tuple[int, Node]] = [(0, root)]

            i = 0
            while i < len(q):
                if q[i][1].children:
                    q.extend(map(lambda x: (q[i][0] + 1, x), q[i][1].children))
                i += 1

            d = defaultdict(list)
            for t in q:
                d[t[0]].append(t[1])

            return [list(map(lambda x: x.val, t)) for t in d.values()]

File: Algorithms/429_N-ary_Tree_Level_Order_Traversal/Solution.kt
/**
 * Definition for a Node.
 * class Node(var `val`: Int) {
 *     var children: List<Node?> = listOf()
 * }
 */

class Solution {
    fun levelOrder(root: Node?): List<List<Int>> = if (root == null) emptyList() else {
        val list = ArrayList<Pair<Int, Node>>().also { it += Pair(0, root) }

        var i = 0
        while (i < list.size) {
            list += list[i++].run {
                this.second.children.map { Pair(this.first + 1, it!!) }
            }
        }

        list.groupBy(Pair<Int, Node>::first) { it.second.`val` }.values.toList()
    }
}